File: pyforms_web_experimental/static/pyformsjs/ControlMap.js
class ControlMap extends ControlBase {

    ////////////////////////////////////////////////////////////////////////////////
    init_control() {
        super.init_control();
        this.map_objects = {};

        this.jquery_place().replaceWith(
            `<div id='${this.place_id()}' class='field control ControlMap' >
                ${this.properties.label_visible ? `<label for="${this.control_id()}">${this.properties.label}</label>` : ''}
                <div style="min-height: ${this.properties.min_height}px;" id='${this.control_id()}' placeholder='${this.properties.placeholder}' />
            </div>`
        );

        if (this.properties.error)
            this.jquery_place().addClass('error');
        else
            this.jquery_place().removeClass('error');

        if (this.properties.required) this.set_required();


        // Delay the map initialization for rendering proposes.
        setTimeout(() => {
            this.initMap();
        }, 500);
    };

    ////////////////////////////////////////////////////////////////////////////////

    initMap() {
        this.map = L.map(this.control_id()).setView(
            this.properties.center,
            this.properties.zoom
        );
        this.initEditPolygon();
        this.process_commands();
    }

    initEditPolygon() {
        this.editableLayers = new L.FeatureGroup();
        this.map.addLayer(this.editableLayers);

        let options = {
            position: 'topleft',
            draw: {
                polyline: this.properties.edit_polyline,
                polygon: this.properties.edit_polygon,
                marker: this.properties.edit_marker,
                circlemarker: this.properties.edit_circlemarker,
                circle: this.properties.edit_circle,
                rectangle: this.properties.edit_rectangle
            }
        };

        if (
            this.properties.edit_polyline ||
            this.properties.edit_polygon ||
            this.properties.edit_marker ||
            this.properties.edit_circlemarker ||
            this.properties.edit_circle ||
            this.properties.edit_rectangle
        ) {
            options.edit = { featureGroup: this.editableLayers, remove: true };
        }

        this.drawControl = new L.Control.Draw(options);
        this.map.addControl(this.drawControl);

        this.map.on(L.Draw.Event.CREATED, e => {
            this.editableLayers.addLayer(e.layer);
            this.basewidget.fire_event(this.name, 'update_control_event');
        });

        this.map.on(L.Draw.Event.EDITED, e => {
            this.basewidget.fire_event(this.name, 'update_control_event');
        });

        this.map.on(L.Draw.Event.DELETED, e => {
            this.basewidget.fire_event(this.name, 'update_control_event');
        });
    }

    process_commands() {

        if (this.properties.add_markers) {
            this.properties.add_markers.forEach((m, idx) => {
                const marker = L.marker(m.coordinate, m);
                this.editableLayers.addLayer(marker);
            });
        }

        if (this.properties.commands) {
            this.properties.commands.forEach(c => {                
                switch (c.command) {
                    case 'clearLayers':
                        for (const key in this.map_objects)
                            if (key.startsWith('layer-')) {
                                this.map_objects[key].removeFrom(this.map);
                                delete this.map_objects[key];
                            }
                        break;
                    case 'fitBounds':
                        this.map.fitBounds(c.bounds)
                        break;
                    case 'addLayer':
                        this.map_objects[`layer-${c.name}`] = L.tileLayer(c.url, c.options).addTo(this.map);
                        break;
                    case 'removeLayer':
                        this.map_objects[`layer-${c.name}`].removeFrom(this.map);
                        delete this.map_objects[`layer-${c.name}`];
                        break;
                    case 'setZIndex':
                        this.map_objects[`layer-${c.name}`].setZIndex(c.z_index);
                        break;
                    case 'setOpacity':
                        this.map_objects[`layer-${c.name}`].setOpacity(c.opacity);
                        break;

                    case 'clearMarkers':
                        for (const key in this.map_objects)
                            if (key.startsWith('marker-')) {
                                this.map_objects[key].removeFrom(this.map);
                                delete this.map_objects[key];
                            }
                        break;
                    case 'addMarker':
                        
                        if(c.iconUrl !== undefined || c.iconUrl !== null){
                            console.debug('addMarker', c);
                            c.options.icon = L.icon({
                                iconUrl: c.iconUrl,
                                iconSize: [25, 25]
                            });
                        }
                        console.debug('-addMarker', c);
                        this.map_objects[`marker-${c.name}`] = L.marker(c.coordinate, c.options).addTo(this.map);
                        break;
                    case 'removeMarker':
                        this.map_objects[`marker-${c.name}`].removeFrom(this.map);
                        delete this.map_objects[`marker-${c.name}`];
                        break;


                    case 'addEditableMarker':
                        const marker = L.marker(c.coordinate, c.options);
                        this.editableLayers.addLayer(marker);
                        this.map_objects[`editable-marker-${c.name}`] = marker;
                        break;
                    case 'addPolyline':
                        this.map_objects[`polyline-${c.name}`] = L.polyline(c.coordinates, c.options).addTo(this.map);
                        break;
                    case 'removePolyline':
                        this.map_objects[`polyline-${c.name}`].removeFrom(this.map);
                        delete this.map_objects[`polyline-${c.name}`];
                        break;
                    case 'clearPolylines':
                        for (const key in this.map_objects)
                            if (key.startsWith('polyline-')) {
                                this.map_objects[key].removeFrom(this.map);
                                delete this.map_objects[key];
                            }
                        break;
                    case 'addPolygon':
                        this.map_objects[`polygon-${c.name}`] = L.polygon(c.coordinates, c.options).addTo(this.map);
                        break;
                    case 'removePolygon':
                        this.map_objects[`polygon-${c.name}`].removeFrom(this.map);
                        delete this.map_objects[`polygon-${c.name}`];
                        break;

                    case 'addEditablePolygon':
                        const poly = L.polygon(c.coordinates, c.options);
                        this.editableLayers.addLayer(poly);
                        this.map_objects[`editable-polygon-${c.name}`] = poly;
                        break;

                        ß
                }
            });
        }
        this.properties.commands = [];

    }

    serialize() {
        const polygons = [];
        const markers = [];
        this.map.eachLayer(l => {
            if (l instanceof L.Polygon) {
                polygons.push(l.toGeoJSON())
            } else if (l instanceof L.Marker && l.options && l.options.pane === "markerPane") {
                markers.push(l.toGeoJSON())
            }
        });

        this.properties.polygons = polygons;
        this.properties.markers = markers;

        return super.serialize();
    }

    deserialize(data) {
        super.deserialize(data);
        if (this.map) {
            this.process_commands();
        }
    }

}

File: pyforms_web_experimental/controls/control_map.py
import simplejson

from pyforms_web.controls.control_base import ControlBase


class ControlMap(ControlBase):

    def __init__(self, *args, **kwargs):
        """
        :param function on_enter_event: Event called when the Enter key is pressed.
        """
        super().__init__(*args, **kwargs)

        self.on_enter_event = kwargs.get('on_enter_event', self.on_enter_event)

        self._edit_polyline = kwargs.get('edit_polyline', False)
        self._edit_polygon = kwargs.get('edit_polygon', False)
        self._edit_marker = kwargs.get('edit_marker', False)
        self._edit_circlemarker = kwargs.get('edit_circlemarker', False)
        self._edit_circle = kwargs.get('edit_circle', False)
        self._edit_rectangle = kwargs.get('edit_rectangle', False)

        self._min_height = kwargs.get('min_height', 180);

        self.center = kwargs.get('center', [51.505, -0.09])
        self.zoom = kwargs.get('zoom', 10)

        self._layers = {}

        self._commands = []

        for layer in self.value:
            self.add_layer(layer['name'], layer['url'], options=layer['options'])

    def init_form(self):
        """
        Initialize the form
        """
        return """new ControlMap('{0}', {1})""".format(
            self._name,
            simplejson.dumps(self.serialize())
        )
    
    @property
    def layers(self):
        """
        Return the layers
        """
        return self._layers

    def on_enter_event(self):
        """
        Event called when the Enter key is pressed
        """
        pass

    def clear_layers(self):
        """
        Clear all layers from the map
        """
        self._commands.append({'command': 'clearLayers'})
        self._layers = {}
        self.mark_to_update_client()

    def fit_bounds(self, bounds):
        """
        Fit the map to the given bounds
        """
        self._commands.append({'command': 'fitBounds', 'bounds': bounds})
        self.mark_to_update_client()

    def add_layer(self, name, url, options=None):
        """
        Add a layer to the map
        """
        self._commands.append({'command': 'addLayer', 'url': url, 'name': name, 'options': options})
        self._layers[name] = url
        self.mark_to_update_client()

    def remove_layer(self, name):
        """
        Remove a layer from the map
        """
        self._commands.append({'command': 'removeLayer', 'name': name})
        self._layers.pop(name)
        self.mark_to_update_client()

    def set_z_index(self, name, z_index):
        """
        Set the z-index of a layer
        """
        self._commands.append({'command': 'setZIndex', 'name': name, 'z_index': z_index})
        self.mark_to_update_client()

    def set_layer_opacity(self, name, opacity):
        """
        Set the opacity of a layer
        """
        self._commands.append({'command': 'setOpacity', 'name': name, 'opacity': opacity})
        self.mark_to_update_client()



    def clear_markers(self):
        """
        Clear a markers from the map
        """
        self._commands.append({'command': 'clearMarkers'})
        self.mark_to_update_client()

    def add_marker(self, name, lat, long, icon=None, **kwargs):
        """
        Add a marker to the map
        """
        self._commands.append({
            'command': 'addMarker', 'coordinate': [lat, long], 'name': name,
            'iconUrl': icon,
            'options': kwargs
            })
        self.mark_to_update_client()

    def remove_marker(self, name):
        """
        Remove a marker from the map
        """
        self._commands.append({'command': 'removeMarker', 'name': name})
        self.mark_to_update_client()



    def add_editable_marker(self, name, lat, long, **kwargs):
        """
        Add a marker to the map
        """
        self._commands.append({'command': 'addEditableMarker', 'coordinate': [lat, long], 'name': name, 'options': kwargs})
        self.mark_to_update_client()

    def add_polyline(self, name, coordinates, **kwargs):
        """
        Add a polyline to the map
        """
        self._commands.append({'command': 'addPolyline', 'coordinates': coordinates, 'name': name, 'options': kwargs})
        self.mark_to_update_client()

    def remove_polyline(self, name):
        """
        Remove a polyline from the map
        """
        self._commands.append({'command': 'removePolyline', 'name': name})
        self.mark_to_update_client()

    def clear_polylines(self):
        """
        Clear a polylines from the map
        """
        self._commands.append({'command': 'clearPolylines'})
        self.mark_to_update_client()

    def add_polygon(self, name, coordinates, **kwargs):
        """
        Add a polygon to the map
        """
        self._commands.append({'command': 'addPolygon', 'coordinates': coordinates, 'name': name, 'options': kwargs})
        self.mark_to_update_client()

    def remove_polygon(self, name):
        """
        Remove a polygon from the map
        """
        self._commands.append({'command': 'removePolygon', 'name': name})
        self.mark_to_update_client()

   
    def add_editable_polygon(self, name, coordinates, **kwargs):
        """
        Add a editable polygon to the map
        """
        self._commands.append({'command': 'addEditablePolygon', 'coordinates': coordinates, 'name': name, 'options': kwargs})
        self.mark_to_update_client()

    def deserialize(self, properties):
        """
        Deserialize the properties
        """

        self.polygons = []
        for p in properties.get('polygons', []):
            poly = [(lat, long) for (long, lat) in p['geometry']['coordinates'][0]]
            self.polygons.append(poly)

        self.markers = []
        for m in properties.get('markers', []):
            coord = m['geometry']['coordinates']
            self.markers.append([coord[1], coord[0]])

        return super().deserialize(properties)

    def serialize(self):
        """
        Serialize the properties
        """

        res = super().serialize()
        res.update({
            'center': self.center,
            'zoom': self.zoom,
            'min_height': self._min_height,
            'commands': self._commands,
            'edit_polyline': self._edit_polyline,
            'edit_polygon': self._edit_polygon,
            'edit_marker': self._edit_marker,
            'edit_circlemarker': self._edit_circlemarker,
            'edit_circle': self._edit_circle,
            'edit_rectangle': self._edit_rectangle,
        })
        self._commands = []

        return res
